notion-server/src/ethereum/mod.rs

```diff
@@ -1,25 +1,122 @@
+#![allow(unused)]
+
+use async_std::sync::{channel, Receiver, Sender};
+
+use snafu::{OptionExt, Snafu};
+
 use std::collections::HashMap;
 
+/// Shorthand for result types returned from the Simulation simulation.
+pub type Result<V, E = Error> = std::result::Result<V, E>;
+
+/// Errors arising from the simulation.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    /// Operation was cancelled because the simulation is shutting down.
+    Terminated,
+}
+
+#[derive(Debug)]
+enum Operation {
+    CreateExecutionEnvironment(args::CreateExecutionEnvironment, Sender<EeIndex>),
+    CreateShardChain(args::CreateShardChain, Sender<u32>),
+    CreateShardBlock(args::CreateShardBlock, Sender<u32>),
+    GetShardBlock(args::GetShardBlock, Sender<ShardBlock>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Handle(Sender<Operation>);
+
+impl Handle {
+    pub async fn create_shard_chain(&self, arg: args::CreateShardChain) -> Result<u32> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::CreateShardChain(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn create_shard_block(&self, arg: args::CreateShardBlock) -> Result<u32> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::CreateShardBlock(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn create_execution_environment(
+        &self,
+        arg: args::CreateExecutionEnvironment,
+    ) -> Result<EeIndex> {
+        let (sender, receiver) = channel(1);
+
+        self.0
+            .send(Operation::CreateExecutionEnvironment(arg, sender))
+            .await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn shard_block(&self, arg: args::GetShardBlock) -> Result<ShardBlock> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::GetShardBlock(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+}
+
 #[derive(Debug)]
-pub struct EthereumSimulation {
+pub struct Simulation {
     beacon_chain: BeaconChain,
     shard_chains: Vec<ShardChain>,
+    receiver: Receiver<Operation>,
 }
 
-impl EthereumSimulation {
-    pub fn new() -> Self {
-        Self {
+impl Simulation {
+    pub fn new() -> (Self, Handle) {
+        let (sender, receiver) = channel(1);
+
+        let sim = Self {
             beacon_chain: BeaconChain::new(),
             shard_chains: Vec::new(),
+            receiver,
+        };
+
+        let handle = Handle(sender);
+
+        (sim, handle)
+    }
+
+    pub async fn run(mut self) -> Result<()> {
+        eprintln!("Simulation Running");
```

Comment:
Why standard error?

Reply:
As opposed to standard out, or why am I outputting at all?

I tend to use stderr for logs, since it isn't output. If you were to pipe the output of the server into another command for further processing, you wouldn't want this line to appear.

As to why I'm printing at all, I'm lazy and didn't set up a proper logging framework yet.